FEATURE: main page

ScortarIrina opened the pull request:
- main page (with all the posts sorted in descending order) is partially implemented
- users and posts can be added (users just to test the functionality)

ScortarIrina committed:
FEATURE: main page

- main page (with all the posts sorted in descending order) is partially implemented
- users and posts can be added (users just to test the functionality)

Laurcons commented on src/app/post-list/post-list.component.ts:
```diff
@@ -0,0 +1,45 @@
+        });
+    }
+
+  formatPublicationDate(dateString: string): string {
```
This better works implemented as a pipe. If you supply a function, like this, Angular has no choice but to consider it impure and evaluate it on every render, even if the parameter `dateString` is the same and wouldn't normally require a render.

You can try to implement such a pipe yourself, or just add something like [time-ago-pipe - npm](https://www.npmjs.com/package/time-ago-pipe). This "time ago" feature is worth its own package.

Laurcons commented on src/app/services/post.service.ts:
```diff
@@ -0,0 +1,24 @@
+    providedIn: 'root'
+})
+export class PostService {
+    private postsUrl = 'http://localhost:8080/post';
```
This way of providing the URL is not appropriate. The accepted way of providing a `baseUrl` into the HttpClient (that is, `http://localhost:8080`) is by using a HTTP interceptor.

We will need the interceptor anyway to provide authentication capabilities. It's okay if it is added in a future PR.

Laurcons commented on src/app/add-post/add-post.component.ts:
```diff
@@ -0,0 +1,40 @@
+
+  async onSubmit() {
+    this.post.publicationDate = new Date().toISOString();
+    this.postService.addPost(this.post).subscribe(
```
We're writing `async/await` JavaScript, so stuff like `.subscribe` needs to be avoided.

In the PostService, we should wrap the Observable that gets returned using `lastValueFrom()`. See an example in a comment in the PostService.

Laurcons commented on src/app/services/post.service.ts:
```diff
@@ -0,0 +1,24 @@
+
+    getPosts(): Observable<any[]> {
+        const url = `${this.postsUrl}/all`;
+        return this.http.get<any[]>(url);
```
```suggestion
        return lastValueFrom(this.http.get<any[]>(url));
```
This Observable returns only once, so it can safely be converted to a Promise using `lastValueFrom` (if I recall correctly the function name). Promises can be awaited and lead to cleaner "linear" code.

Additionally, please provide a proper type instead of `any`. This will provide type-checking and autocompletion all the way down to the template. This might be an oversight since the `addPost` function below is properly typed.

You can also omit the function return type. It is clear from the `this.http.get<>` call what gets returned (and TypeScript is perfectly fine with it too).

Laurcons commented on src/app/services/user.service.ts:
```diff
@@ -0,0 +1,23 @@
+@Injectable({
+  providedIn: 'root'
+})
+export class UserService {
```
This class is very similar to UserService.

To provide some help with such repeating code, an abstract `CrudService` could be created with basic CRUD operations.

I don't really see this as utterly necessary and can be done in a future PR if wanted.

Laurcons commented on src/app/user-list/user-list.component.ts:
```diff
@@ -0,0 +1,19 @@
+  styleUrls: ['./user-list.component.scss']
+})
+export class UserListComponent implements OnInit {
+  users: any[] = [];
```
Please properly type this array

Laurcons commented on src/app/add-post/add-post.component.scss:
```diff
@@ -0,0 +1,10 @@
+.form-group {
```
We want to move to Angular Material but I have to say that Bootstrap does not need to be styled like this. Everything that this rule declares can be applied with the classes `d-flex`, `align-center` and `p-3` (or similar).

Modifying Boostrap classes is not advised anyway, even at component-level.

Laurcons reviewed (changes requested):
Interesting choice of writing tests. Unfortunately, I don't think our team can get any value from these tests (and we agreed we wouldn't do unit testing).

We have also agreed to use Angular Material for UI styling. I should have configured that in the project a long time ago, so this one's on me. If you want, you can try rewriting the UI using Angular Material, or you can let me do it in a future PR.

Laurcons commented:
@davidalexandru1370 has made #4 which adds Angular Material on the project, and has overall fewer changes to the main branch. I would advise waiting for his PR to merge before syncing `main` into your branch and fixing conflicts (namely, theming and the UserService).

PS. Doing it the other way around would lead to too many conflicts

ScortarIrina committed:
FIX: changes from PR review

- removed AddUser form and menu entry (useless)
- properly typed the arrays
- removed deprecated .subscribe()
- created pipe for time formatting the time-ago for posts

ScortarIrina committed:
FIX: changes from PR review

- removed AddUser form and menu entry (useless)
- properly typed the arrays
- removed deprecated .subscribe()
- created pipe for time formatting the time-ago for posts
- added some Angular Material

ScortarIrina committed:
Merge branch 'main' into task/main-page

# Conflicts:
#	angular.json
#	package-lock.json
#	package.json
#	src/app/app-routing.module.ts
#	src/app/app.component.html
#	src/app/app.module.ts
#	src/app/services/user.service.ts
#	src/styles.scss

ScortarIrina committed:
task: integrated login into main page

- a user needs to login in order to use the app

ScortarIrina committed:
task: integrated login into main page

- a user needs to login in order to use the app
- each post is shown with its corresponding author
- logout button
